src/backend/distributed/operations/shard_split.c

```diff
@@ -1346,107 +1340,125 @@ NonBlockingShardSplit(SplitOperation splitOperation,
 		 * Subscriber flow starts from here.
 		 * Populate 'ShardSplitSubscriberMetadata' for subscription management.
 		 */
-		List *shardSplitSubscribersMetadataList =
+		List *logicalRepTargetList =
 			PopulateShardSplitSubscriptionsMetadataList(
-				shardSplitHashMapForPublication, replicationSlotInfoList);
+				publicationInfoHash, replicationSlotInfoList,
+				shardGroupSplitIntervalListList, workersForPlacementList);
+
+		HTAB *groupedLogicalRepTargetsHash = CreateGroupedLogicalRepTargetsHash(
+			logicalRepTargetList);
 
 		/* Create connections to the target nodes */
-		List *targetNodeConnectionList = CreateTargetNodeConnectionsForShardSplit(
-			shardSplitSubscribersMetadataList,
-			connectionFlags,
+		CreateGroupedLogicalRepTargetsConnections(
+			groupedLogicalRepTargetsHash,
 			superUser, databaseName);
 
-		/* 7) Create copies of template replication slot */
-		char *templateSlotName = ShardSplitTemplateReplicationSlotName(
-			shardIntervalToSplit->shardId);
-		CreateReplicationSlots(sourceConnection, templateSlotName,
-							   shardSplitSubscribersMetadataList);
+		char *logicalRepDecoderPlugin = "citus";
+
+		/*
+		 * 6) Create replication slots and keep track of their snapshot.
+		 */
+		char *snapshot = CreateReplicationSlots(
+			sourceConnection,
+			sourceReplicationConnection,
+			logicalRepTargetList,
+			logicalRepDecoderPlugin);
+
+		/*
+		 * 7) Create subscriptions. This isn't strictly needed yet at this
+		 * stage, but this way we error out quickly if it fails.
+		 */
+		CreateSubscriptions(
+			sourceConnection,
+			databaseName,
+			logicalRepTargetList);
+
+		/* 8) Do snapshotted Copy */
+		DoSplitCopy(sourceShardToCopyNode, sourceColocatedShardIntervalList,
+					shardGroupSplitIntervalListList, workersForPlacementList,
+					snapshot);
 
-		/* 8) Create subscriptions on target nodes */
-		CreateShardSplitSubscriptions(targetNodeConnectionList,
-									  shardSplitSubscribersMetadataList,
-									  sourceShardToCopyNode,
-									  superUser,
-									  databaseName);
 
 		/* Used for testing */
 		ConflictOnlyWithIsolationTesting();
```

Comment:
not due to this PR, but kind of wondering whether this call happens a bit early, there might be a bunch of operations below that could accidentally block writes before we get to BlockWritesToShardList.